### DISCUSSION TITLE
X11+Wayland+KDE+Gnome watcher on Rust

### DISCUSSION THREAD

**Discussion by 2e3s** [ActivityWatch/activitywatch#874] [General]
Hi!
Thank you for your program! I've been a fairly long term user of your program, but I missed or didn't like a few things.
- I'm a forever KDE user, and recently switched to Wayland. XWayland doesn't report all active windows.
- I also used to switch to X11 sometimes.
- Sometimes my data should not be reported, or reported differently.
- Subjectively, Python based mess is discouraging comparing to 1 binary.

Thanks to inspiration from your new Rust server and some productivity issues of mine, I've finally decided to dedicate a few weeks to implement everything (no productivity issues for that).

Here is [X11+Wayland+KDE+Gnome watcher on Rust](https://github.com/2e3s/awatcher). I don't need all of that, but 1 environment takes just 1 evening, maybe it'll help somebody. I could also add Windows, although that would take a few evenings for setting things up.

Now I need a build script to package the server for convenience, and probably a tray option, but maybe you would be interesting in incorporating such a watcher instead?

**Comment by ErikBjare** [comment]
Wow, very nice! @johan-bjareholt is probably interested :)

Would you mind creating a PR adding it to the docs? It's this page: https://github.com/ActivityWatch/docs/blob/master/src/watchers.rst

> Now I need a build script to package the server for convenience, and probably a tray option, but maybe you would be interesting in incorporating such a watcher instead?

If you feel up for it, take a look at https://github.com/ActivityWatch/aw-tauri which is a WIP Rust-only ActivityWatch distribution built with Tauri. Your watcher would make a perfect addition to complete the bundle! (it currently has no watchers)

**Comment by 2e3s** [comment]
> Would you mind creating a PR adding it to the docs?

Where would my watcher go? "Other watchers"? This section has non-window watchers though...
https://docs.activitywatch.net/en/latest/watchers.html

> which is a WIP Rust-only ActivityWatch distribution built with Tauri

I considered Tauri too for bundling. Actually, I've spent the past day playing with it and quite amazingly I was able to build a working desktop application of ActivityWatch within such a limited time frame: https://github.com/2e3s/tawatcher (there is an alpha release which is a working executable, at least it works for me). It's my aforementioned Linux watchers + aw-server-rust, I took the snippet from your link.

It's a bit heavy (like hell) taking 93 mb RAM. I haven't used the Tauri's server for aw-webui because aw-webui uses a hardcoded URL for API client which is the same URL as HTML URL. Instead it redirects to the page which is hosted by aw-server, i.e. localhost:5600. It would be easy to fix if necessary in [ActivityWatch/aw-webui](https://github.com/ActivityWatch/aw-webui/). Also, header and footer are weird in this context, maybe they could be removed.

<details>
<summary>Screenshot</summary>

![image](https://user-images.githubusercontent.com/4003445/235386696-3ead7689-09c3-46c5-af85-ebf3b3c76d19.png)

</details>

Because of the heaviness, I still ponder on using a tray for [2e3s/awatcher](https://github.com/2e3s/awatcher), at least as an optional cargo feature. I has actually done it already with `ksni` crate, it's very easy and all is native. But maybe Tauri-driven application can get hacks, like unloading WebKit when it's deemed unnecessary. It's nevertheless a bit strange as a mere wrapper (and I don't see much opportunity to integrate Rust code and the existing web UI).

**Comment by 2e3s** [comment]
So, instead of Tauri, I've made a bundle where aw-webui is simply compiled into the binary (same RAM usage though). Plus tray icon to open webui. Just a single executable, more lightweight than the webkit base stuff https://github.com/2e3s/awatcher/releases/tag/v0.0.2
